lib/ui/painting/paint.cc

```diff
@@ -98,7 +98,14 @@ const SkPaint* Paint::paint(SkPaint& paint) const {
       Shader* decoded = tonic::DartConverter<Shader*>::FromDart(shader);
       auto sampling =
           ImageFilter::SamplingFromIndex(uint_data[kFilterQualityIndex]);
-      paint.setShader(decoded->shader(sampling)->skia_object());
+      auto color_source = decoded->shader(sampling);
+      // TODO(dnfield): Remove this restriction.
+      // This currently is only used by paragraph code. Once SkParagraph does
+      // not need to take an SkPaint, we won't be restricted in this way because
+      // we will not need to access the shader on the UI task runner.
+      if (color_source->owning_context() != DlImage::OwningContext::kRaster) {
```

Comment:
I'm wondering where SkParagraph accesses this. AFAIK, they store the SkPaint object and they use it to render to our SkCanvas adapter, but they never actually do anything with its contents. Is this a theoretical concern, or was there a case where we ended up accessing the ImageSource from the wrong thread?

Reply:
It's access here: https://github.com/flutter/engine/blob/5f92d3668c7a172be23df13eebd24ab7e7df29ee/lib/ui/text/paragraph_builder.cc#L468-L482

Which gets passed along to SkPAragraph eventually

Reply:
It is copying the paint into a local field, but it isn't accessing the contents of the paint.

Reply:
Those get held and then sent to us when the paragraph is drawn into a ui.Picture, and then we hold on to it until we eventually use it to render on the raster thread. Or at least that is my understanding of how we interact with SkParagraph.

Reply:
We shouldn't need to create an `SkPaint` at all here once @jason-simmons' work is done, AFAIU.